File: Day 11/permissions.py
from typing import Set

from fastapi import Depends, HTTPException

from auth import get_current_user
from models import User

ROLE_PERMISSIONS: dict[str, Set[str]] = {
    "admin": {
        "users:read",
        "users:write",
        "users:delete",
        "tasks:read",
        "tasks:write",
        "tasks:delete",
    },
    "editor": {
        "tasks:read",
        "tasks:write",
        "tasks:delete",
        "users:read",
    },
    "viewer": {
        "tasks:read",
        "users:read",
    },
}

DEFAULT_ROLE = "editor"


def get_permissions(role: str) -> Set[str]:
    return ROLE_PERMISSIONS.get(role, set())


def require_permission(permission: str):
    async def checker(current_user: User = Depends(get_current_user)) -> User:
        perms = get_permissions(current_user.role)
        if permission not in perms:
            raise HTTPException(
                status_code=403,
                detail=f"Missing permission: {permission}",
            )
        return current_user

    return checker

File: Day 11/frontend/src/permissions.ts
export type UserRole = 'admin' | 'editor' | 'viewer'

export type Permission =
  | 'users:read'
  | 'users:write'
  | 'users:delete'
  | 'tasks:read'
  | 'tasks:write'
  | 'tasks:delete'

const ROLE_PERMISSIONS: Record<UserRole, ReadonlySet<Permission>> = {
  admin: new Set([
    'users:read',
    'users:write',
    'users:delete',
    'tasks:read',
    'tasks:write',
    'tasks:delete',
  ]),
  editor: new Set(['tasks:read', 'tasks:write', 'tasks:delete', 'users:read']),
  viewer: new Set(['tasks:read', 'users:read']),
}

export function getPermissions(role: string): Set<Permission> {
  return new Set(ROLE_PERMISSIONS[role as UserRole] ?? [])
}

export function hasPermission(role: string, permission: Permission): boolean {
  return getPermissions(role).has(permission)
}

export function formatRole(role: string): string {
  return role.charAt(0).toUpperCase() + role.slice(1)
}

export function isAdmin(role: string): boolean {
  return role === 'admin'
}
